src/v/compression/async_stream_zstd.cc

```diff
@@ -0,0 +1,262 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "compression/async_stream_zstd.h"
+
+#include "bytes/bytes.h"
+#include "bytes/details/io_allocation_size.h"
+#include "compression/logger.h"
+#include "likely.h"
+#include "units.h"
+#include "vassert.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/coroutine/maybe_yield.hh>
+
+#include <fmt/format.h>
+
+#include <array>
+#include <zstd.h>
+#include <zstd_errors.h>
+
+namespace compression {
+[[gnu::cold]] static void throw_zstd_err(size_t rc) {
+    if (rc == ZSTD_error_memory_allocation) {
+        ss::throw_with_backtrace<std::bad_alloc>();
+    } else {
+        ss::throw_with_backtrace<std::runtime_error>(
+          fmt::format("ZSTD error:{}", ZSTD_getErrorName(rc)));
+    }
+}
+
+static void throw_if_error(size_t rc) {
+    if (unlikely(ZSTD_isError(rc))) {
+        throw_zstd_err(rc);
+    }
+}
+
+static constexpr size_t default_decompression_size = 2_MiB;
+static constexpr int default_compression_level = 3;
+static constexpr int default_contexts = 2;
+
+static thread_local std::unique_ptr<async_stream_zstd> zstd_instance;
+
+void initialize_async_stream_zstd(size_t decompression_size) {
+    if (zstd_instance) {
+        vassert(
+          decompression_size <= zstd_instance->decompression_size(),
+          "async_stream_zstd initialized multiple times with "
+          "decompression_size > zstd_instance->decompression_size()");
+
+        return;
+    }
+
+    zstd_instance = std::make_unique<async_stream_zstd>(
+      decompression_size, default_contexts);
+}
+
+async_stream_zstd& async_stream_zstd_instance() {
+    if (unlikely(!zstd_instance)) {
+        initialize_async_stream_zstd(default_decompression_size);
+    }
+
+    return *zstd_instance;
+}
+
+async_stream_zstd::static_cctx::static_cctx(int compression_level) {
+    _c_buffer = ss::temporary_buffer<char>(ZSTD_CStreamOutSize());
+    auto c_workspace_size = ZSTD_estimateCStreamSize(compression_level);
+    // zstd requires alignment for the buffer below.
+    _c_workspace = ss::allocate_aligned_buffer<char>(c_workspace_size, 8);
+
+    vassert(
+      _c_workspace,
+      "Failed to allocate zstd workspace of size {}",
+      c_workspace_size);
+
+    _compress_ctx = ZSTD_initStaticCCtx(_c_workspace.get(), c_workspace_size);
+
+    throw_if_error(ZSTD_CCtx_setParameter(
+      _compress_ctx, ZSTD_c_compressionLevel, default_compression_level));
+}
+
+async_stream_zstd::static_cctx::static_cctx(static_cctx&& ctx) noexcept
+  : _compress_ctx(ctx._compress_ctx)
+  , _c_workspace(std::move(ctx._c_workspace))
+  , _c_buffer(std::move(ctx._c_buffer)) {
+    ctx._compress_ctx = nullptr;
```

Comment:
Nothing to change but it's a shame you have to write these. The default one _almost_ does the right thing, except for setting the `_compress_ctx` to null. It's almost like you'd want a dumb pointer wrapper which did this for you on move.